### PR TITLE
auth-user

### PR DESCRIPTION
**Pull request by ptran068**
Login, Create, logout

### CODE REVIEW

File: src/users/models.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+from django.contrib.auth.models import AbstractUser
+from .managers import CustomUserManager
+import uuid
+# from files.models import File
+
+class CustomUser(AbstractUser):
+    username = None
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False, unique=True)
+    email = models.EmailField(('email address'), unique=True)
+    # avatar = models.ForeignKey(File, on_delete=models.CASCADE, blank=True, null=True)
```
Comment (prd-anh-nguyen) [inline src/users/models.py]: if don't use it. please remove

File: src/users/models.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    phone = models.CharField(blank=True, null=True, max_length=15)
+    is_admin = models.BooleanField(default=False)
+    USERNAME_FIELD = 'email'
```
Comment (prd-anh-nguyen) [inline src/users/models.py:11]: where is you use it?

File: src/users/models.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from django.db import models
+from django.contrib.auth.models import AbstractUser
+from .managers import CustomUserManager
+import uuid
+# from files.models import File
```
Comment (prd-anh-nguyen) [inline src/users/models.py]: remove

File: src/users/forms.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+from django import forms
+
+# class CreateUser(forms.ModelForm):
```
Comment (prd-anh-nguyen) [inline src/users/forms.py]: ?

File: src/middlewares/authentication.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        except (jwt.DecodeError, model.DoesNotExist):
+            raise exceptions.AuthenticationFailed('Invalid token')
+        except jwt.ExpiredSignatureError:
+            raise exceptions.AuthenticationFailed('Token has expired')
+        if not user.is_active:
+            raise exceptions.AuthenticationFailed('User inactive or deleted')
+        return (user, payload)
+
+
+class SessionAuthentication(BaseAuthentication):
```
Comment (prd-anh-nguyen) [inline src/middlewares/authentication.py:49]: do you want to override SessionAuthentication class?
what different with origin class?

File: src/users/models.py (hunk 2 above)
Comment (ptran068) [reply src/users/models.py:11]: default authentication system of django rest_framwork and superuser

File: src/middlewares/authentication.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    def authenticate(self, request):
+        auth = get_authorization_header(request).split()
+        if not auth or auth[0].lower() != b'token':
+            return None
+
+        if len(auth) == 1:
+            msg = 'Invalid token header. No credentials provided.'
+            raise exceptions.AuthenticationFailed(msg)
+        elif len(auth) > 2:
```
Comment (prd-anh-nguyen) [inline src/middlewares/authentication.py]: what is the format of correct auth? I don't know when u check len(auth) must = 2?

File: src/users/managers.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+from django.contrib.auth.base_user import BaseUserManager
+
+class CustomUserManager(BaseUserManager):
```
Comment (prd-anh-nguyen) [inline src/users/managers.py:3]: what do you think about `Manager` responsibility?

File: src/middlewares/authentication.py (hunk 5 above)
Comment (ptran068) [reply src/middlewares/authentication.py:49]: No it the same, but i'd like to show it to my team members in order to make it clearly, i also remove away

File: src/users/managers.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    def create_user(self, email, password, **extra_fields):
+        if not email:
+            raise ValueError(_('The Email must be set'))
+        email = self.normalize_email(email)
+        user = self.model(email=email, **extra_fields)
+        user.set_password(password)
+        user.save()
+        return user
+
+    def create_superuser(self, email, password, **extra_fields):
```
Comment (prd-anh-nguyen) [inline src/users/managers.py:13]: I didn't see where this function to use

File: src/middlewares/authentication.py (hunk 6 above)
Comment (ptran068) [reply src/middlewares/authentication.py]: The correct auth is composed of token name space and  real token, EX: token asdsadsdsdsa12s1ad, so i did it to ensure that length is equal 2.

File: src/users/managers.py (hunk 8 above)
Comment (ptran068) [reply src/users/managers.py:13]: if u dont override usernam by email,The system will use usename to generate admin account, but i'd like to use email for my system

File: src/users/managers.py (hunk 7 above)
Comment (ptran068) [reply src/users/managers.py:3]: If u dont satisfy about default field of User system, u can use it to additional field and custom anything what u want. I think so

File: src/middlewares/authentication.py (hunk 6 above)
Comment (prd-anh-nguyen) [reply src/middlewares/authentication.py]: please use regex to verify token's format

File: src/users/managers.py (hunk 7 above)
Comment (prd-anh-nguyen) [reply src/users/managers.py:3]: `A Manager is the interface through which database query operations are provided to Django models` 
Manager `database query` not a function to update object. Please research and use `Services` to do it

File: src/middlewares/authentication.py (hunk 6 above)
Comment (ptran068) [reply src/middlewares/authentication.py]: i have realized the function get_authorization_header in django restframwork make sure the token from header is valid

File: src/users/managers.py (hunk 7 above)
Comment (ptran068) [reply src/users/managers.py:3]: oh sorry, i dont understand what u say before @@.I just create a manager.py to  shortcut  model file.and i override it in model file

Comment (ptran068) [reply src/users/managers.py:3]: objects = CustomUserManager()

Comment (ptran068) [reply src/users/managers.py:3]: image

**Review by ntdo2506 — APPROVED** [review]

**Review by datnguyen0126 — APPROVED** [review]

**Review by prd-anh-nguyen — APPROVED** [review]

**Pull request merged**